include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,1286 @@
+
+    // Perform a read.
+    // ref_load(ref, rgn, res);
+    // Definition: assign abstract value stored in rgn to res
+    //  rgn: a region variable used in object domain
+    //  res: a register variable used in base domain
+    // The transformer is sound for ref_load required the followings:
+    //  1. The type of region variable is consistent with the type of register
+    //  E.g. if region represents integer, the register must be an integer type
+    //  2. The reference ref is not evaluated as null pointer.
+    //  3. The object domain of the abstract object is either:
+    //     object domain is Top (i.e. the object domain is not found in odi map)
+    //  or object domain exists
+    //  3.1 if object domain is Top, the overapproximate res value as top
+    //  3.2 if object domain exists, extract interval for rgn in object domain
+    //    assign res := interval into base domain
+    //
+    //  post condition: the odi map not changed, the base domain is updated
+
+    ERROR_IF_NOT_REGION(rgn, __LINE__);
+    ERROR_IF_ARRAY_REGION(rgn, __LINE__);
+    ERROR_IF_NOT_REF(ref, __LINE__);
+    if ((rgn.get_type().is_bool_region() && !res.get_type().is_bool()) ||
+        (rgn.get_type().is_integer_region() && !res.get_type().is_integer()) ||
+        (rgn.get_type().is_real_region() && !res.get_type().is_real()) ||
+        (rgn.get_type().is_reference_region() &&
+         !res.get_type().is_reference())) {
+      CRAB_ERROR(domain_name(), "::ref_load: type of lhs ", res, " (",
+                 res.get_type(), ") is not compatible with region ", rgn, " (",
+                 rgn.get_type(), ")");
+    }
+    // At this point, the requirement for 1 is satisfied
+
+    if (is_bottom()) {
+      return;
+    }
+
+    if (is_null_ref(ref).is_true()) {
+      CRAB_LOG("object-load", CRAB_WARN(domain_name(), "::ref_load: reference ",
+                                        ref, " is null."););
+      // set_to_bottom();
+      crab::CrabStats::count(domain_name() +
+                             ".count.ref_load.skipped.null_reference");
+      m_base_dom -= res;
+      return;
+    }
+    // At this point, the requirement for 2 is satisfied
+
+    // The odi must exist for each region after instrisic calls
+    variable_t rep = get_odi(rgn);
+
+    // retrieve an abstract object info
+    auto obj_info = m_obj_info_env.at(rep);
+
+    const small_range &num_refs = obj_info.refcount_val();
+
+    // In crab IR, the number of references cannot be zero
+    //  if ref_load access a not null reference.
+    // So zero case should not exist
+    assert(!num_refs.is_zero());
+
+    // read from odi map
+    const field_abstract_domain_t *obj_dom_ref = m_odi_map.find(rep);
+    if (!obj_dom_ref) {  // not found in odi map, means object domain is top
+      m_base_dom -= res; // forget res means res == top
+      // the requirment 3.1 is satisfied
+    } else {
+      field_abstract_domain_t *obj_dom_ref_no_const =
```

Comment:
There are two issues with returning an interval. First, the underlying domain might not keep any interval representation. That's why the returned interval is by value. Second,  the method `interval_t operator[](...)` is not marked as `const` because some domains (e.g., octagons) needs to apply normalization before producing the interval. I could have a second `const` version that cannot normalize.

Reply:
A C++ convention is to provide a method called `at()` that is same as `operator[]` but const.

The domain should not expose its internal normalization through `const`. It should hide it using `mutable`. 
At the very least, you want to provide a const `at` method, that, for example, does `const cast` under the hood.